cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/output/MultiOutputCommitter.java

```diff
@@ -122,8 +112,67 @@ public boolean isRecoverySupported() {
 
   @Override
   public void recoverTask(TaskAttemptContext taskContext) throws IOException {
+    delegateInParallel((name, delegate) -> {
+      TaskAttemptContext namedContext = MultiOutputFormat.getNamedTaskContext(taskContext, name);
+      delegate.recoverTask(namedContext);
+    });
+  }
+
+  private void delegateInParallel(DelegateFunction delegateFunction) throws IOException {
+    ExecutorService executorService = Executors.newFixedThreadPool(delegates.size(), new ThreadFactory() {
```

Comment:
Use a ThreadFactoryBuilder instead to avoid holder a reference in the finalizer (see https://github.com/cdapio/cdap/pull/12567)